### PR TITLE
nixos/test-driver: use pythons logging module

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg test ferm

File: nixos/lib/test-driver/test-driver.py [hunk 1]
```diff
@@ -88,22 +89,25 @@
...
+logging.basicConfig(format="%(message)s")
+logger = logging.getLogger("test-driver")
+logger.setLevel(logging.INFO)
 
-def eprint(*args: object, **kwargs: Any) -> None:
-    print(*args, file=sys.stderr, **kwargs)
+
+class MachineLogAdapter(logging.LoggerAdapter):
+    def process(self, msg: str, kwargs: Any) -> Any:
+        return "[%s] %s" % (self.extra["machine"], msg), kwargs
```
Comment (tfc) [inline nixos/lib/test-driver/test-driver.py:102]:
```suggestion
        return f"[{self.extra["machine"]}] {msg}", kwargs
```

File: nixos/lib/test-driver/test-driver.py [hunk 2]
```diff
@@ -88,22 +89,25 @@
...
 
 
 def make_command(args: list) -> str:
     return " ".join(map(shlex.quote, (map(str, args))))
 
 
 def create_vlan(vlan_nr: str) -> Tuple[str, str, "subprocess.Popen[bytes]", Any]:
-    global log
-    log.log("starting VDE switch for network {}".format(vlan_nr))
+    logger.info("starting VDE switch for network {}".format(vlan_nr))
```
Comment (tfc) [inline nixos/lib/test-driver/test-driver.py:110]:
```suggestion
    logger.info(f"starting VDE switch for network {vlan_nr}"
```

File: nixos/lib/test-driver/test-driver.py [hunk 3]
```diff
@@ -381,16 +313,16 @@ def systemctl(self, q: str, user: Optional[str] = None) -> Tuple[int, str]:
         return self.execute("systemctl {}".format(q))
 
     def require_unit_state(self, unit: str, require_state: str = "active") -> None:
-        with self.nested(
+        logger.info(
```
Comment (tfc) [inline nixos/lib/test-driver/test-driver.py:316]: `self.`?

File: nixos/lib/test-driver/test-driver.py [hunk 4]
```diff
@@ -566,15 +492,12 @@ def screenshot(self, filename: str) -> None:
...
-            "making screenshot {}".format(filename),
-            {"image": os.path.basename(filename)},
-        ):
-            self.send_monitor_command("screendump {}".format(tmp))
-            ret = subprocess.run("pnmtopng {} > {}".format(tmp, filename), shell=True)
-            os.unlink(tmp)
-            if ret.returncode != 0:
-                raise Exception("Cannot convert screenshot")
+        self.logger.info(f"making screenshot {filename}")
+        self.send_monitor_command("screendump {}".format(tmp))
```
Comment (tfc) [inline nixos/lib/test-driver/test-driver.py:496]: i know that you basically just changed the indentation of this code after removing the nesting that we don't support any longer. but can you please at the same time homogenize a bit the formatting to use more `f"screendump {tmp}"` style?

**Review by tfc — CHANGES_REQUESTED** [review]

**Comment by JJJollyjim** [comment]
Haha, I just opened github to open the same PR

**Comment by JJJollyjim** [comment]
Here's mine for comparison: https://github.com/JJJollyjim/nixpkgs/tree/python-logging

File: nixos/lib/test-driver/test-driver.py (hunk 4 above)
Comment (JJJollyjim) [reply nixos/lib/test-driver/test-driver.py:496]: Alternatively, `logging` has its own formatting which might make sense to use if we're committing to using it (`self.logger.info("making screenshot %s", filename)`)

Comment (tfc) [reply nixos/lib/test-driver/test-driver.py:496]: I am generally not against this, but the automatic type selection using `{}` looks favorable to me in contrast to `%s` style formatting.

Comment (JJJollyjim) [reply nixos/lib/test-driver/test-driver.py:496]: I believe `%s` will always act the same as `{}`:
```
>>> "%s" % 5.6
'5.6'
>>>
```

(unlike C where it causes horrifying undefined behaviour :D)

**Comment by JJJollyjim** [comment]
btw, as-is this will break a few tests which were depending on log functions:

2020-08-25-222342_838x934_scrot

**Comment by JJJollyjim** [comment]
Personally, I'd like to see automatic (streaming) output to the logger from `.succeed` and related functions, which would remove the need for most of these log statements (and generally make test development easier in some cases)

File: nixos/lib/test-driver/test-driver.py (hunk 4 above)
Comment (Mic92) [reply nixos/lib/test-driver/test-driver.py:496]: No, the PR is already quite big. This is another refactoring.

**Comment by Mic92** [comment]
@GrahamcOfBorg test nsd

**Comment by Mic92** [comment]
I re-added the `log()` function.

**Comment by JJJollyjim** [comment]
lgtm :)

**Comment by flokli** [comment]
Thanks for this! :-)

File: nixos/lib/test-driver/test-driver.py [hunk 5]
```diff
@@ -292,14 +228,6 @@ def create_startcommand(args: Dict[str, str]) -> str:
     def is_up(self) -> bool:
         return self.booted and self.connected
 
-    def log(self, msg: str) -> None:
-        self.logger.log(msg, {"machine": self.name})
-
-    def nested(self, msg: str, attrs: Dict[str, str] = {}) -> _GeneratorContextManager:
-        my_attrs = {"machine": self.name}
-        my_attrs.update(attrs)
-        return self.logger.nested(msg, my_attrs)
```
Comment (andersk) [inline nixos/lib/test-driver/test-driver.py:301]: Removing `Machine.nested` has broken a number of tests that used it, including the channel-blocking chromium test:

```console
$ git grep '\.nested' nixos/tests
nixos/tests/chromium.nix:        with machine.nested("Creating a new Chromium window"):
nixos/tests/chromium.nix:        with machine.nested("Waiting for new Chromium window to appear"):
nixos/tests/chromium.nix:        with machine.nested(description):
nixos/tests/dhparams.nix:      with machine.nested("switch to generation ${toString gen}"):
nixos/tests/dhparams.nix:        with machine.nested(f"check bit size of {path}"):
nixos/tests/initrd-network-ssh/default.nix:    with client.nested("waiting for SSH server to come up"):
nixos/tests/taskserver.nix:            with client.nested(f"initialize client for user {user}"):
nixos/tests/taskserver.nix:                with client.nested("importing taskwarrior configuration"):
nixos/tests/taskserver.nix:        with server.nested("(re-)add imperative user bar"):
nixos/tests/virtualbox.nix:      with machine.nested("Checking for privilege escalation"):
nixos/tests/virtualbox.nix:    with machine.nested("Checking for privilege escalation"):
```

**Comment by andersk** [comment]
This was reverted for now in #96703 to fix #96699.

**Pull request merged**